python/ray/serve/tests/test_http_state.py

```diff
@@ -3,172 +3,517 @@
+            raise Exception("Never be healthy")
+
+    proxy_state = _create_http_proxy_state(proxy_actor_class=NewMockHTTPProxyActor)
+
+    # Continuously trigger update. The status should change from STARTING to HEALTHY
+    # when ready.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+    first_check_time = proxy_state._last_health_check_time
+
+    # Ensure the proxy status is HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
+
+    # Continuously trigger update and status should change from HEALTHY to UNHEALTHY.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.UNHEALTHY,
+    )
+    # Ensure the proxy status is still UNHEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.UNHEALTHY
+
+    # Ensure the check time have changed since the last update
+    assert first_check_time != proxy_state._last_health_check_time
+
+    # Ensure _consecutive_health_check_failures is correct
+    assert proxy_state._consecutive_health_check_failures == 3
+
+
+@patch("ray.serve._private.http_state.DEFAULT_HEALTH_CHECK_TIMEOUT_S", 0.1)
+@patch("ray.serve._private.http_state.PROXY_HEALTH_CHECK_PERIOD_S", 0.1)
+def test_http_proxy_state_update_healthy_check_health_always_timeout():
+    """Test calling update method on HTTPProxyState when the proxy state is HEALTHY and
+    when the ready call always timed out.
+
+    The proxy state started with STARTING. After update is called and ready call
+    succeeded, the status will change to HEALTHY. After the next few check_health calls
+    takes very long time to finished, the status will eventually change to UNHEALTHY
+    after all retries have exhausted.
+    """
+
+    @ray.remote(num_cpus=0)
+    class NewMockHTTPProxyActor:
+        async def ready(self):
+            return json.dumps(["mock_worker_id", "mock_log_file_path"])
+
+        async def check_health(self):
+            await asyncio.sleep(100)
 
-    ray.shutdown()
+    proxy_state = _create_http_proxy_state(proxy_actor_class=NewMockHTTPProxyActor)
+
+    # Continuously trigger update. The status should change from STARTING to HEALTHY
+    # when ready.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+    first_check_time = proxy_state._last_health_check_time
+
+    # Ensure the proxy status is HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
+
+    # Continuously trigger update and status continue to be UNHEALTHY.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.UNHEALTHY,
+    )
+    # Ensure the proxy status is still UNHEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.UNHEALTHY
```

Comment:
Just want to be as explicit as possible so others who might have not have seen this code would know what this is checking. But let me know if you feel strongly we shouldn't do redundant checks like this, I can drop all such checks

Reply:
in other places we rely only on the `wait_for_condition`, so suggest to remove the redundant `assert`

Reply:
Talked with Ed, since this is a common knowledge that `wait_for_condition()` will also do assertion and is used everywhere, will just change all those explicit assertions. Same goes to the other comment 🙂